### PR TITLE
Fix IOS bug in media Element when using Popup

### LINKED ISSUE #1421: [BUG] MediaElement Crash in CollectionView On IOS
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

When placing a MediaElement inside a CollectioView, the app will crash when run on iOS, but not Android. See the error below:

ObjCRuntime.ObjCException: Objective-C exception thrown.  Name: UIViewControllerHierarchyInconsistency Reason: child view controller:<AVPlayerViewController: 0x7f7a03092c00> should have parent view controller:<Microsoft_Maui_Controls_Handlers_Items_ReorderableItemsViewController_1: 0x7f79fcf64af0> but actual parent is:<Microsoft_Maui_Controls_Platform_Compatibility_ShellFlyoutRenderer: 0x7f79fcf35510>
Native stack trace:
	0   CoreFoundation                      0x00007ff8004288ab __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007ff80004dba3 objc_exception_throw + 48
	2   CoreFoundation                      0x00007ff800428789 -[NSException initWithCoder:] + 0
	3   UIKitCore                           0x000000011fc0295c -[UIView(Hierarchy) _associatedViewControllerForwardsAppearanceCallbacks:performHierarchyCheck:isRoot:] + 299
	4   UIKitCore                           0x000000011fc031ba -[UIView(Hierarchy) _willMoveToWindow:withAncestorView:] + 249
	5   UIKitCore                           0x000000011fc0408a _makeSubTreePerformSelector + 661
	6   UIKitCore                           0x000000011fc040d4 _makeSubTreePerformSelector + 735
	7   UIKitCore                           0x000000011fc040d4 _makeSubTreePerformSelector + 735
	8   UIKitCore                           0x000000011fc13b14 -[UIView(Internal) _addSubview:positioned:relativeTo:] + 491
	9   UIKitCore                           0x000000011fb988ac -[UIScrollView _addContentSubview:atBack:] + 566
	10  UIKitCore                           0x000000011eadcefd -[UICollectionView _addControlledSubview:atZIndex:forced:initialAttributes:] + 970
	11  UIKitCore                           0x000000011eae1d9c -[UICollectionView _createPreparedCellForItemAtIndexPath:withLayoutAttributes:applyAttributes:isFocused:notify:] + 1637
	12  UIKitCore                           0x000000011eae1731 -[UICollectionView _createPreparedCellForItemAtIndexPath:withLayoutAttributes:applyAttributes:] + 31
	13  UIKitCore                           0x000000011eaebd47 -[UICollectionView _createVisibleViewsForSingleCategoryAttributes:limitCreation:fadeForBoundsChange:] + 1731
	14  UIKitCore                           0x000000011eaec24b -[UICollectionView _createVisibleViewsForAttributes:fadeForBoundsChange:notifyLayoutForVisibleCellsPass:] + 376
	15  UIKitCore                           0x000000011eae9fad -[UICollectionView _updateVisibleCellsNow:] + 2331
	16  UIKitCore                           0x000000011eaf0bc4 -[UICollectionView layoutSubviews] + 375
	17  UIKitCore                           0x000000011fc1a59f -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 2305
	18  QuartzCore                          0x00007ff80898b02c _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 526
	19  QuartzCore                          0x00007ff8089965b9 _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 65
	20  QuartzCore                          0x00007ff8088a7b0c _ZN2CA7Context18commit_transactionEPNS_11TransactionEdPd + 626
	21  QuartzCore                          0x00007ff8088df69b _ZN2CA11Transaction6commitEv + 735
	22  UIKitCore                           0x000000011f5e622b _UIApplicationFlushRunLoopCATransactionIfTooLate + 70
	23  UIKitCore                           0x000000011f6b43be __processEventQueue + 9443
	24  UIKitCore                           0x000000011f6aa2a1 __eventFetcherSourceCallback + 272
	25  CoreFoundation                      0x00007ff800387fe5 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	26  CoreFoundation                      0x00007ff800387f24 __CFRunLoopDoSource0 + 157
	27  CoreFoundation                      0x00007ff800387721 __CFRunLoopDoSources0 + 212
	28  CoreFoundation                      0x00007ff800381e23 __CFRunLoopRun + 927
	29  CoreFoundation                      0x00007ff8003816a7 CFRunLoopRunSpecific + 560
	30  GraphicsServices                    0x00007ff809cb128a GSEventRunModal + 139
	31  UIKitCore                           0x000000011f5e7ad3 -[UIApplication _run] + 994
	32  UIKitCore                           0x000000011f5ec9ef UIApplicationMain + 123
	33  libxamarin-dotnet-debug.dylib       0x000000010b76c64a xamarin_UIApplicationMain + 58
	34  libmonosgen-2.0.dylib               0x000000010bf7a5a9 do_icall + 345
	35  libmonosgen-2.0.dylib               0x000000010bf790b3 do_icall_wrapper + 291
	36  libmonosgen-2.0.dylib               0x000000010bf6b4af interp_exec_method + 3295
	37  libmonosgen-2.0.dylib               0x000000010bf68e03 interp_runtime_invoke + 259
	38  libmonosgen-2.0.dylib               0x000000010bd8e0e8 mono_runtime_invoke_checked + 136
	39  libmonosgen-2.0.dylib               0x000000010bd960fc do_exec_main_checked + 92
	40  libmonosgen-2.0.dylib               0x000000010bec4ce2 mono_jit_exec + 354
	41  libxamarin-dotnet-debug.dylib       0x000000010b7b0c77 xamarin_main + 1927
	42  CommunityToolkit.Maui.Sample        0x000000010aeff6f4 main + 52
	43  dyld                                0x000000010b44a2bf start_sim + 10
	44  ???                                 0x000000010cd0541f 0x0 + 4509946911

### Expected Behavior

Users can see multiple videos at a time.

### Steps To Reproduce

1. Create new .net Maui app with .net 7
2. Add CommunityToolkit.Maui.MediaElement (version 2.0.0) Nuget package
3. In MauiProgram.cs, add .UseMauiCommunityToolkitMediaElement()
4. In MainPage.xaml, replace ScrollView content with a MediaElement
5. Confirm it runs correctly.
6. Replace the ScrollView with a CollectionView. Set a generic ItemSource array and set the MediaElement as the ItemTemplate.
7. Confirm it crashes on iOS.

### Link to public reproduction project repository

https://github.com/vtmonilgandhi/MediaElementIssue

### Environment

```markdown
- .NET MAUI CommunityToolkit: 5.3.0
- OS: 16.2, 17.0
- .NET MAUI: 7.0
```

### Anything else?

_No response_

### LINKED ISSUE #1556: MediaElement causes crash when used inside a popup on iOS only
---
**Issue moved from dotnet/maui#18952**
- Please respond to @Swodah.

---

_From @Swodah on Wednesday, November 22, 2023 1:19:07 PM_

### Description

A page CheckInOutLandingPage
call static method that takes the page as a variable: 
`await CheckInOutHelper.ShowCheckOutSuccess(this);`
method does the following:
```
  var dialog = new AnimDialog
  {
      IconAnimation = "warning.json",
      Title = MauiProgram.translator.GetText("CheckinFailedCode"),
      OkButton = ShowRetry == true ? MauiProgram.translator.GetText("Retry") : null,
      CancelButton = MauiProgram.translator.GetText("Back"),
      AudioFile = MediaSource.FromResource("CheckInOutFail.mp3"),
      CanBeDismissedByTappingOutsideOfPopup = false
  };

  var result = await parent.ShowPopupAsync(dialog);
  return result;
```

Note AnimDialog is a page that inherits from CommunityToolkit.Maui. popup

On Windows and android it plays the popup without issues.
On iOS it causes a crash.
Previously we got no crash but the page was broken.

### Steps to Reproduce

_No response_

### Link to public reproduction project repository

_No response_

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

7.0.100

### Affected platforms

iOS

### Affected platform versions

ios 17.0 ipad simulator and iPadOS 17.1.1 physical device

### Did you find any workaround?

_No response_

### Relevant log output

```shell
the following stack trace:

   at ObjCRuntime.Runtime.ThrowException(IntPtr gchandle) in /Users/builder/azdo/_work/1/s/xamarin-macios/src/ObjCRuntime/Runtime.cs:line 2594
   at UIKit.UIApplication.UIApplicationMain(Int32 argc, String[] argv, IntPtr principalClassName, IntPtr delegateClassName) in /Users/builder/azdo/_work/1/s/xamarin-macios/src/UIKit/UIApplication.cs:line 60
   at UIKit.UIApplication.Main(String[] args, Type principalClass, Type delegateClass) in /Users/builder/azdo/_work/1/s/xamarin-macios/src/UIKit/UIApplication.cs:line 94
   at UMS_APP_MAUI.Program.Main(String[] args) in D:\Projects\Applab_MAUI\UMS_APP_MAUI\Platforms\iOS\Program.cs:line 12
```
and this message:
`ObjCRuntime.ObjCException: 'Objective-C exception thrown.  Name: UIViewControllerHierarchyInconsistency Reason: child view controller:<AVPlayerViewController: 0x7fe4b5ddc000> should have parent view controller:<Microsoft_Maui_Platform_PageViewController: 0x7fe4b83321c0> but actual parent is:<Microsoft_Maui_Platform_PageViewController: 0x7fe4b3b742c0>
Native stack trace:
	0   CoreFoundation                      0x0000000111a7428d __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x000000012362c894 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000111a7416b -[NSException initWithCoder:] + 0
	3   UIKitCore                           0x000000013cfb7bb5 -[UIView(Hierarchy) _associatedViewControllerForwardsAppearanceCallbacks:performHierarchyCheck:isRoot:] + 287
	4   UIKitCore                           0x000000013cfb84bd -[UIView(Hierarchy) _willMoveToWindow:withAncestorView:] + 330
	5   UIKitCore                           0x000000013cfb94c0 _makeSubTreePerformSelector + 655
	6   UIKitCore                           0x000000013cfb9502 _makeSubTreePerformSelector + 721
	7   UIKitCore                           0x000000013cfb9502 _makeSubTreePerformSelector + 721
	8   UIKitCore                           0x000000013cfb9502 _makeSubTreePerformSelector + 721
	9   UIKitCore                           0x000000013cfb9502 _makeSubTreePerformSelector + 721
	10  UIKitCore                           0x000000013cfb9502 _makeSubTreePerformSelector + 721
	11  UIKitCore                           0x000000013cfb9502 _makeSubTreePerformSelector + 721
	12  UIKitCore                           0x000000013cfb9502 _makeSubTreePerformSelector + 721
	13  UIKitCore                           0x000000013cfb9502 _makeSubTreePerformSelector + 721
	14  UIKitCore                           0x000000013cfc9c41 -[UIView(Internal) _addSubview:positioned:relativeTo:] + 524
	15  UIKitCore                           0x000000013bf68291 __67-[UIViewControllerBuiltinTransitionViewAnimator animateTransition:]_block_invoke_2 + 46
	16  UIKitCore                           0x000000013cfc2ac4 +[UIView(Animation) performWithoutAnimation:] + 84
	17  UIKitCore                           0x000000013bf6821b __67-[UIViewControllerBuiltinTransitionViewAnimator animateTransition:]_block_invoke + 115
	18  UIKitCore                           0x000000013cfc447f +[UIView _setupAnimationWithDuration:delay:view:options:factory:animations:start:animationStateGenerator:completion:] + 593
	19  UIKitCore                           0x000000013cfc4ad1 +[UIView(UIViewAnimationWithBlocks) animateWithDuration:delay:options:animations:completion:] + 25
	20  UIKitCore                           0x000000013bf67fba -[UIViewControllerBuiltinTransitionViewAnimator animateTransition:] + 3624
	21  UIKitCore                           0x000000013bd7f45e -[_UIModernPopoverAnimationController animateTransition:] + 606
	22  UIKitCore                           0x000000013bf70a2d ___UIViewControllerTransitioningRunCustomTransition_block_invoke_2 + 59
	23  UIKitCore                           0x000000013c107975 +[UIKeyboardSceneDelegate _pinInputViewsForKeyboardSceneDelegate:onBehalfOfResponder:duringBlock:] + 100
	24  UIKitCore                           0x000000013bf709b8 ___UIViewControllerTransitioningRunCustomTransition_block_invoke.649 + 178
	25  UIKitCore                           0x000000013cfc2810 +[UIView(Animation) _setAlongsideAnimations:toRunByEndOfBlock:] + 176
	26  UIKitCore                           0x000000013bf70839 _UIViewControllerTransitioningRunCustomTransition + 577
	27  UIKitCore                           0x000000013be007a1 __56-[UIPresentationController runTransitionForCurrentState]_block_invoke.430 + 2431
	28  UIKitCore                           0x000000013cf5799c -[_UIAfterCACommitBlock run] + 55
	29  UIKitCore                           0x000000013cf57e82 -[_UIAfterCACommitQueue flush] + 189
	30  UIKitCore                           0x000000013c913b5f _runAfterCACommitDeferredBlocks + 784
	31  UIKitCore                           0x000000013c902a05 _cleanUpAfterCAFlushAndRunDeferredBlocks + 96
	32  UIKitCore                           0x000000013c9377f8 _afterCACommitHandler + 58
	33  CoreFoundation                      0x00000001119cfd94 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	34  CoreFoundation                      0x00000001119ca72a __CFRunLoopDoObservers + 534
	35  CoreFoundation                      0x00000001119cac72 __CFRunLoopRun + 1158
	36  CoreFoundation                      0x00000001119ca409 CFRunLoopRunSpecific + 557
	37  GraphicsServices                    0x000000012a81d187 GSEventRunModal + 137
	38  UIKitCore                           0x000000013c9043a2 -[UIApplication _run] + 972
	39  UIKitCore                           0x000000013c908e10 UIApplicationMain + 123
	40  libxamarin-dotnet-debug.dylib       0x000000010edb74da xamarin_UIApplicationMain + 58
	41  libmonosgen-2.0.dylib               0x000000010f555bc5 do_icall + 341
	42  libmonosgen-2.0.dylib               0x000000010f554357 do_icall_wrapper + 295
	43  libmonosgen-2.0.dylib               0x000000010f545076 mono_interp_exec_method + 3990
	44  libmonosgen-2.0.dylib               0x000000010f5427e3 interp_runtime_invoke + 259
	45  libmonosgen-2.0.dylib               0x000000010f628908 mono_runtime_invoke_checked + 136
	46  libmonosgen-2.0.dylib               0x000000010f6300db mono_runtime_exec_main_checked + 107
	47  libmonosgen-2.0.dylib               0x000000010f494f52 mono_jit_exec + 354
	48  libxamarin-dotnet-debug.dylib       0x000000010edfbbba xamarin_main + 1898
	49  UMS_APP_MAUI                        0x00000001046345d4 main + 68
	50  dyld                                0x000000010d6023ee start_sim + 10
	51  ???                                 0x0000000204d3e3a6 0x0 + 8670929830
'`
```

### PR DESCRIPTION
**Pull request by ne0rrmatrix**
- Bug fix for Media Element trying to add controls to wrong View Controller.

 ### Description of Change ###
- Adds checks to make sure video controls do not get added to wrong View Controller. This works if using Shell, Page, Navigation Page, Tabbed Page, Popup, CollectionView, and CarouselView

 ### Linked Issues ###
 - Fixes #1556 
 - Fixes #1421

 ### PR Checklist ###
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls 

 ### Additional information ###

 Fixes CTD when using media element in a Popup. The incorrect Parent ViewController is being used when adding controls to parent view controller. A sample for popup has been added to samples for media element. This document has been edited since written to better explain the changes and what the fix does.

### CODE REVIEW

**Comment by bijington** [comment]
@ne0rrmatrix thank you for submitting this. Sorry to ask this but is there an easy way to test this fix? I'm guessing there isn't anything in the sample that can be tested?

**Comment by ne0rrmatrix** [comment]
@bijington you can load this PR in the quoted issues example repo.

**Comment by bijington** [comment]
> @bijington you can load this PR in the quoted issues example repo. This fix requires targeting dotnet 8.0.x.

OK thanks. Do you think there is any value in us introducing a similar example in the sample application? I'm happy to assist getting it in as part of the testing.

**Comment by brminnick** [comment]
> Do you think there is any value in us introducing a similar example in the sample application?

Absolutely! Always 💯 

Anytime we fix a bug, we should add an example of its reproduction to the Sample App. This helps us verify the bug fix and ensures that we won't have a regression when testing a future PR.

**Comment by ne0rrmatrix** [comment]
@brminnick I will work on adding reproduction to the sample app this week? I assume adding with PR is a good idea? Or should it be seperate? I am thinking separate because if we end up reverting PR the example goes away?

**Comment by brminnick** [comment]
Thanks! Yes, please add it to this PR

**Comment by ne0rrmatrix** [comment]
@brminnick I rewrote some View Controller logic for IOS. This change allows the controls to actually work. I still have a bunch of testing to do so please wait to review. I added a Popup with mediaElement inside it to the Media element page as a button on the far right called Popup in the sample app.

**Comment by ne0rrmatrix** [comment]
Have done some testing and fixed a bug with Navigation Page logic. I still have to test Flyout Page and Tabbed Page. Once I confirm they are working I will tag someone to review.

**Comment by ne0rrmatrix** [comment]
Ok I have tested against shell, navigation Page, Content Page, and Tabbed Page. Does anyone have any other suggested tests?

**Comment by ne0rrmatrix** [comment]
Fix for collection view controller when using Carousel View has been added. Anyone who looks at this for review please indicate whether the approach I use is appropriate. There is now at least one other PR out to fix same issue. I am not sure which approach is appropriate and what the maintainers feel is best to move forward. I am happy either way.

File: src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs [hunk 1]
```diff
@@ -13,16 +15,15 @@ public class MauiMediaElement : UIView
 	/// Initializes a new instance of the <see cref="MauiMediaElement"/> class.
 	/// </summary>
 	/// <param name="playerViewController">The <see cref="AVPlayerViewController"/> that acts as the platform media player.</param>
-	/// <param name="parentViewController">The <see cref="UIViewController"/> that acts as the parent for <paramref name="playerViewController"/>.</param>
 	/// <exception cref="NullReferenceException">Thrown when <paramref name="playerViewController"/><c>.View</c> is <see langword="null"/>.</exception>
-	public MauiMediaElement(AVPlayerViewController playerViewController, UIViewController? parentViewController)
+	public MauiMediaElement(AVPlayerViewController playerViewController)
```
Comment (pictos) [inline src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:19]: just to point out that this's a breaking change

Comment (ne0rrmatrix) [reply src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:19]: @pictos  can you give me an example?

Comment (ne0rrmatrix) [reply src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:19]: There might be a better solution: #1623 
@pictos if you can look at that PR and get back to me if that is a better solution that does not cause breaking changes? Someone else is working on a related problem that may make this PR moot.

Comment (pictos) [reply src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:19]: @ne0rrmatrix I'm not saying this is wrong or bad... Just point out that's a breaking change, so if this PR gets merged we will need to release a major version. What you can do is preserve the old ctor and mark it with the Obsolete attribute and not show it on Intellisense.

Comment (ne0rrmatrix) [reply src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:19]: Ok i will restore old ctor and mark obsolete attribute.

Comment (ne0rrmatrix) [reply src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:56]: @pictos  I reverted ctor and marked old ctor as ```[Obsolete]``` Hope I did it right. Please review and suggest any changes.

File: src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs [hunk 2]
```diff
@@ -15,14 +17,15 @@ public class MauiMediaElement : UIView
...
+	[Obsolete("Depracated CTOR. UIViewController? parentViewController is obsolete")]
 	public MauiMediaElement(AVPlayerViewController playerViewController, UIViewController? parentViewController)
 	{
 		ArgumentNullException.ThrowIfNull(playerViewController.View);
 		playerViewController.View.Frame = Bounds;
 
 #if IOS16_0_OR_GREATER || MACCATALYST16_1_OR_GREATER
 		// On iOS 16+ and macOS 13+ the AVPlayerViewController has to be added to a parent ViewController, otherwise the transport controls won't be displayed.
-		var viewController = parentViewController ?? WindowStateManager.Default.GetCurrentUIViewController();
+		var viewController = WindowStateManager.Default.GetCurrentUIViewController();
```
Comment (pictos) [inline src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:28]: Can you revert the changes on this ctor? This is still a breaking change.

Comment (ne0rrmatrix) [reply src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:30]: Yes I will do that now.

Comment (ne0rrmatrix) [reply src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs]: @pictos Updated as requested

**Review by pictos — DISMISSED** [review]

**Review by pictos — APPROVED** [review]
